opnode/l1/source.go

```diff
@@ -0,0 +1,370 @@
+func (s *Source) FetchAllTransactions(ctx context.Context, window []eth.BlockID) ([]types.Transactions, error) {
+	// list of transaction lists
+	allTxLists := make([]types.Transactions, len(window))
+
+	var blockRequests []rpc.BatchElem
+	var requestIndices []int
+
+	for i := 0; i < len(window); i++ {
+		// if we are shifting the window by 1 block at a time, most of the results should already be in the cache.
+		txs, ok := s.transactionsCache.Get(window[i].Hash)
+		if ok {
+			allTxLists[i] = txs.(types.Transactions)
+		} else {
+			blockRequests = append(blockRequests, rpc.BatchElem{
+				Method: "eth_getBlockByHash",
+				Args:   []interface{}{window[i].Hash, true}, // request block including transactions list
+				Result: new(rpcBlock),
+				Error:  nil,
+			})
+			requestIndices = append(requestIndices, i) // remember the block index this request corresponds to
+		}
+	}
+
+	if len(blockRequests) > 0 {
+		if err := s.batchCall(ctx, blockRequests); err != nil {
+			return nil, err
+		}
+	}
+
+	// try to cache everything we have before halting on the results with errors
+	for i := 0; i < len(blockRequests); i++ {
+		if blockRequests[i].Error == nil {
+			info, txs, err := blockRequests[i].Result.(*rpcBlock).Info(s.trustRPC)
+			if err != nil {
+				return nil, fmt.Errorf("bad block data for block %s: %v", blockRequests[i].Args[0], err)
+			}
+			s.headersCache.Add(info.hash, info)
+			s.transactionsCache.Add(info.hash, txs)
+			allTxLists[requestIndices[i]] = txs
+		}
+	}
+
+	for i := 0; i < len(blockRequests); i++ {
+		if blockRequests[i].Error != nil {
+			return nil, fmt.Errorf("failed to retrieve transactions of block %s in batch of %d blocks: %v", window[i], len(blockRequests), blockRequests[i].Error)
+		}
+	}
+
+	return allTxLists, nil
+}
+
+func (s *Source) L1HeadBlockRef(ctx context.Context) (eth.L1BlockRef, error) {
+	head, err := s.InfoHead(ctx)
+	if err != nil {
+		return eth.L1BlockRef{}, fmt.Errorf("failed to fetch head header: %v", err)
+	}
+	return head.BlockRef(), nil
+}
+
+func (s *Source) L1BlockRefByNumber(ctx context.Context, l1Num uint64) (eth.L1BlockRef, error) {
+	head, err := s.InfoByNumber(ctx, l1Num)
+	if err != nil {
+		return eth.L1BlockRef{}, fmt.Errorf("failed to fetch header by num %d: %v", l1Num, err)
+	}
+	return head.BlockRef(), nil
+}
+
+func (s *Source) L1BlockRefByHash(ctx context.Context, hash common.Hash) (eth.L1BlockRef, error) {
+	block, err := s.InfoByHash(ctx, hash)
+	if err != nil {
+		return eth.L1BlockRef{}, fmt.Errorf("failed to fetch header by hash %v: %w", hash, err)
+	}
+	return block.BlockRef(), nil
+}
+
+// L1Range returns a range of L1 block beginning just after begin, up to max blocks.
+// This batch-requests all blocks by number in the range at once, and then verifies the consistency
+func (s *Source) L1Range(ctx context.Context, begin eth.BlockID, max uint64) ([]eth.BlockID, error) {
+	headerRequests := make([]rpc.BatchElem, max)
```

Comment:
Couldn't this method also read from the cache?

Reply:
We're fetching blocks by number. The cache is by hash. If the L1 reorgs the keying by number breaks. We only fetch headers here, and do so in large concurrent batches as much as possible, so I think it's fine not to do complex and bug-prone reorg-aware caching.